### PR TITLE
Prettify the sharing page.

### PR DESCRIPTION
**Pull request by blrubenstein**
Default the domain: Would like to change this to use based on environment.
Added non-existant email error message.
Added loading wheel.

### CODE REVIEW

File: ui/src/app/views/workspace-share/component.ts [hunk 1]
```diff
@@ -61,32 +69,72 @@ export class WorkspaceShareComponent implements OnInit {
...
+    this.errorHandlingService.retryApi(
+      this.workspacesService.shareWorkspace(
+        this.workspace.namespace,
+        this.workspace.id, {
+          workspaceEtag: this.workspace.etag,
+          items: this.updateList})).subscribe(
+      (resp: ShareWorkspaceResponse) => {
+        this.workspace.etag = resp.workspaceEtag;
+        this.usersLoading = false;
+        this.workspace.userRoles = this.updateList;
```
Comment (calbach) [inline ui/src/app/views/workspace-share/component.ts]: Seems like this could get hairy if there are concurrent pending requests - `this.updateList` may have changed since you sent it above. Perhaps `updateList` here should be local? Though there are likely many ways to resolve this e.g. disallowing concurrent share requests in some way or else chain the executions of these add/removes.

It should actually always fail to send concurrent requests, since the etag would be stale for one of them, so disabling or queueing these actions probably makes sense (if queueing - lazily reading the current etag before the share request).

File: ui/src/app/views/workspace-share/component.ts [hunk 2]
```diff
@@ -61,32 +69,72 @@ export class WorkspaceShareComponent implements OnInit {
...
+        this.toShare = '';
+        this.input.nativeElement.focus();
+      },
+      (error) => {
+        if (error.status === 400) {
```
Comment (calbach) [inline ui/src/app/views/workspace-share/component.ts]: Not found is 400 and not a 404?

File: ui/src/app/views/workspace-share/component.html [hunk 3]
```diff
@@ -27,19 +27,26 @@ <h4>Add a New Collaborator</h4>
         </clr-dropdown>
         <button type="button" class="btn btn-primary sharing-buttons" (click)="addCollaborator()">Add Collaborator</button>
       </div>
+      <div *ngIf="userNotFound" class="error-message">
```
Comment (calbach) [inline ui/src/app/views/workspace-share/component.html]: opt: a similar reload prompt as I added on the edit page would be nice here

File: ui/src/app/views/workspace-share/component.ts (hunk 1 above)
Comment (blrubenstein) [reply ui/src/app/views/workspace-share/component.ts]: `this.updateList` should be re-created every time `addCollaborator` or `removeCollaborator` is called, so yes, it should be local! Will change. I'm doing it with a separate array to avoid updating the screen until after the API call has actually succeeded. If there are two concurrent requests, one will error with a 409, right? In that case it should go to the error response, rather than the success response. I can work on adding an error case for conflict errors.

File: ui/src/app/views/workspace-share/component.ts (hunk 2 above)
Comment (blrubenstein) [reply ui/src/app/views/workspace-share/component.ts]: Will update on API side (currently throwing a badRequestException?). I'm not actually sure if this should be considered a badRequest or a notFound.

File: ui/src/app/views/workspace-share/component.ts (hunk 1 above)
Comment (johnstill) [reply ui/src/app/views/workspace-share/component.ts]: > disallowing concurrent share requests

> If there are two concurrent requests, one will error with a 409

Better to just cancel any existing outbound requests that have not yet returned (I'd probably use a `switchMap`).  Less work for the servers and less error handling for you to do.

File: ui/src/app/views/workspace-share/component.ts (hunk 2 above)
Comment (johnstill) [reply ui/src/app/views/workspace-share/component.ts]: This should be a 404. 400 should be reserved for actually malformed http.

File: ui/src/app/views/workspace-share/component.ts [hunk 4]
```diff
@@ -61,32 +69,72 @@ export class WorkspaceShareComponent implements OnInit {
...
-    this.workspacesService.shareWorkspace(
-      this.workspace.namespace, this.workspace.id, {
-        workspaceEtag: this.workspace.etag,
-        items: this.workspace.userRoles
-      }).subscribe((resp: ShareWorkspaceResponse) => {
-        this.workspace.etag = resp.workspaceEtag;
+    if (this.inputNotInitialized) {
+      this.inputNotInitialized = false;
+      this.input.nativeElement.addEventListener('keydown', () => {
+        this.userNotFound = false;
```
Comment (johnstill) [inline ui/src/app/views/workspace-share/component.ts]:
```
<input (keydown)="userNotFound = false" /* attrs... */ >
```

Allows you to dispense with the initialization flag and is easier to scan (to my mind).

File: ui/src/app/views/workspace-share/component.ts (hunk 2 above)
Comment (dmohs) [reply ui/src/app/views/workspace-share/component.ts]: I tend to use 404 and 405 to refer specifically to the URL and method, respectively. Otherwise, I'm not sure how to distinguish between, "your request is semantically invalid" and "you fat-fingered the URL." That leaves 400 to be fairly generic. For example, calling myservice/create-user with the payload `{"totallyInvalid": 7}` is a bad request because the payload doesn't have the required information. I would treat a request to /myservice/user/1234 as a 404 if that user doesn't exist because it is a URL. In this case, I'd argue that the request is semantically invalid because you're putting invalid data into the payload (a non-existant account), so I'd go with 400.

To put it another way, how would you distinguish between the following?
```
/api/v1/workspaces/my-namespace/my-workspace/share
{"email": "totally-invalid@researchallofus.org"}
```
and
```
/api/v1/workspaces/my-namespace/my-workspace/sh4re
{"email": "valid-user@researchallofus.org"}
```

Comment (johnstill) [reply ui/src/app/views/workspace-share/component.ts]: having taken a closer look at the [actual api method](https://github.com/all-of-us/workbench/blob/5ad30f86b7e5f7d53aa4b94656d4260edd2a8236/api/src/main/resources/workbench.yaml#L421) I'm going to agree with @dmohs here; if a user isn't found then in this case that indicates bad POST data (a 400).

**Comment by blrubenstein** [comment]
This should be good for another look! I wrote some unit tests, fixed some of our other tests/services, and added the concurrent modification error.

File: ui/src/app/views/workspace-share/component.html [hunk 5]
```diff
@@ -51,6 +51,17 @@ <h5>
     </div>
   </div>
 </div>
+<clr-modal [(clrModalOpen)]="workspaceUpdateConflictError">
+  <h3 class="modal-title">Conflicting update:</h3>
+  <div class="modal-body">
+    Another client has modified this workspace since the beginning of this
+    sharing session. Please reload to avoid overwriting those changes.
```
Comment (dmohs) [inline ui/src/app/views/workspace-share/component.html]: Nice error message here.
